src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs

```diff
@@ -659,6 +646,16 @@ public static void Run(ServiceBase[] services)
             }
             finally
             {
+                // Free the pointer to the name of the service on the unmanaged heap.
+                // We don't need to free the last element in the unmanaged array since the last element have null values (denotes the end of the table).
+                // Entries other than the last entry having a null pointer means that we failed while processing the entries for some reason
+                // and only elements up to this point has been successfully allocated so far (thus have to be deallocated)
+                for (int i = 0; i < services.Length && entries[i].name != IntPtr.Zero; i++)
```

Comment:
Can the serviceName be NULL in some cases? The `entries[i].name != IntPtr.Zero` condition is not going to work if that ever happens.

Reply:
[Per the documents](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshal.stringtohglobaluni?view=netframework-4.8#returns), `Marshal.StringToHGlobalUni` shouldn't return `IntPtr.Zero` unless the string was null. I don't think the string will ever be null (the default value is string.Empty, and the setter disallows null string). In addition, because we cleared the span & assign sequentially from the start to the end, there can't be any non-zero values past any null pointers.

Though, given that `Marshal.FreeHGlobal(IntPtr.Zero)` [is a no-op](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshal.freehglobal?view=netframework-4.8#remarks), it could be worth removing the check altogether. The performance implication wouldn't really matter either, because it is quite an edge case to fail in the loop. Better stay safe than sorry 🙂

Reply:
> the default value is string.Empty

As far as I can tell by looking at the code, the default value is null.

Reply:
It is set in the ctor through the setter of _serviceBase: https://github.com/dotnet/corefx/blob/f546aa9e76c776f2c27e30dfb91c50d8f62da2f8/src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs#L54

Reply:
> it could be worth removing the check altogether. The performance implication wouldn't really matter either, because it is quite an edge case to fail in the loop. Better stay safe than sorry

Agree.